storage/innobase/row/row0import.cc

```diff
@@ -4279,6 +4529,77 @@ static void row_import_autoinc(dict_table_t *table, row_prebuilt_t *prebuilt,
   }
 }
 
+/** Update the virtual column position in SYS_COLUMNS and SYS_VIRTUAL
+@param  table_id   table identifier
+@param  new_pos    position value
+@param  trx        transaction
+@return DB_SUCCESS or error code */
+dberr_t update_vcol_pos(ulint table_id, ulint new_pos, trx_t *trx)
+{
+  pars_info_t *info= pars_info_create();
+  pars_info_add_ull_literal(info, "id", table_id);
+  pars_info_add_int4_literal(info, "old_pos", new_pos - 1);
+  return que_eval_sql(info,
+                      "PROCEDURE UPDATE_VCOL () IS\n"
+                      "BEGIN\n"
+		      "UPDATE SYS_COLUMNS SET POS = POS + 1 "
+		      "WHERE TABLE_ID= :id AND POS = :old_pos;\n"
+		      "UPDATE SYS_VIRTUAL SET POS = POS + 1 "
+		      "WHERE TABLE_ID= :id AND POS = :old_pos;\n"
+		      "END\n;", trx);
+}
+
+/**
+1) Update the position of the columns and
+2) Insert the hidden fts doc id in the sys columns table
+3) Insert the hidden fts doc id in the sys indexes and
+sys_fields table
+@param  table   table to be imported
+@param  fts_pos position of fts doc id column
+@param  trx     transaction
+@return DB_SUCCESS or error code */
+static
+dberr_t innodb_insert_hidden_fts_col(dict_table_t* table,
+                                     ulint  fts_pos,
+                                     trx_t* trx)
+{
+  dict_index_t* fts_idx=
+    dict_table_get_index_on_name(table, FTS_DOC_ID_INDEX_NAME);
+  if (!fts_idx) return DB_ERROR;
+
+  for (ulint new_i= 0; new_i < table->n_v_cols; new_i++)
+  {
+    ulint pos= dict_create_v_col_pos(
+                 table->v_cols[new_i].v_pos,
+		 table->v_cols[new_i].m_col.ind);
+    dberr_t err= update_vcol_pos(table->id, pos, trx);
+    if (err != DB_SUCCESS)
```

Comment:
The idiomatic C++98 way to write this would be as follows:
```c++
if (dberr_t err= update_vcol_pos(table->id, pos, trx))
```
However, why do we invoke `update_vcol_pos()` several times? Does the `WHERE` condition really have to check equality with `pos`?

Is it possible that `pos` are assigned in ascending order in this loop? If yes, there would seem to be a possible duplicate key error  collision on `SYS_VIRTUAL` if there are multiple virtual columns defined on the same base column(s). I see that the test case `innodb.import_hidden_fts` does define two virtual columns that are derived from the same base column, but only one of them is indexed. Could that be the reason why no bug was caught?

Can we add something to ensure that the `pos` are assigned in ascending order? Or compute the exact bound that needs to be passed to `UPDATE SYS_… SET POS=POS+1 WHERE TABLE_ID=:id AND POS>=:old_pos`?